magento2/usr/local/share/magento2/magento_functions.sh

```diff
@@ -34,28 +34,44 @@ function do_magento_create_web_writable_directories() {
 }
 
 function do_magento_frontend_build() {
-  if [ -d "$FRONTEND_INSTALL_DIRECTORY" ]; then
-    mkdir -p pub/static/frontend/
+  do_magento_frontend_build_install
+  do_magento_frontend_build_run
+}
+
+function do_magento_frontend_build_install() {
+  if [ ! -d "$FRONTEND_INSTALL_DIRECTORY" ]; then
+    return
+  fi
 
+  mkdir -p pub/static/frontend/
   if [ -d "pub/static/frontend/" ] && [ "$IS_CHOWN_FORBIDDEN" != 'true' ]; then
-      chown -R "${CODE_OWNER}:${CODE_GROUP}" pub/static/frontend/
-    fi
+    chown -R "${CODE_OWNER}:${CODE_GROUP}" pub/static/frontend/
+  fi
 
-    if [ ! -d "$FRONTEND_INSTALL_DIRECTORY/node_modules" ]; then
-      as_code_owner "npm install" "$FRONTEND_INSTALL_DIRECTORY"
-    fi
-    if [ -z "$GULP_BUILD_THEME_NAME" ]; then
-      as_code_owner "gulp $FRONTEND_BUILD_ACTION" "$FRONTEND_BUILD_DIRECTORY"
-    else
-      as_code_owner "gulp $FRONTEND_BUILD_ACTION --theme='$GULP_BUILD_THEME_NAME'" "$FRONTEND_BUILD_DIRECTORY"
-    fi
+  as_code_owner "npm install" "$FRONTEND_INSTALL_DIRECTORY"
+}
 
-  if [ -d "pub/static/frontend/" ] && [ "$IS_CHOWN_FORBIDDEN" != 'true' ]; then
-      chown -R "${APP_USER}:${APP_GROUP}" pub/static/frontend/
-    fi
+function do_magento_frontend_build_run() {
+  if [ ! -d "$FRONTEND_INSTALL_DIRECTORY" ]; then
+    return
+  fi
+
+  if [ -z "$GULP_BUILD_THEME_NAME" ]; then
+    as_code_owner "gulp $FRONTEND_BUILD_ACTION" "$FRONTEND_BUILD_DIRECTORY"
+  else
+    as_code_owner "gulp $FRONTEND_BUILD_ACTION --theme='$GULP_BUILD_THEME_NAME'" "$FRONTEND_BUILD_DIRECTORY"
   fi
 }
 
+function do_magento_frontend_build() {
+  do_magento_frontend_build_install
+  do_magento_frontend_build_run
+}
+
+function do_magento_frontend_cache_clean() {
+  as_code_owner "npm cache clean --force"
```

Comment:
should we disable this in dev mode to keep things fast?

Reply:
Good point - yes, it would allow a volume for /home/build/.cache/npm later on. I think we should do the same for the composer cache clear too.